sopel/plugins/rules.py

```diff
@@ -1780,6 +1780,10 @@ def from_callable(cls, settings, handler):
             @functools.wraps(handler)
             def execute_handler(bot, trigger):
                 return handler(bot, trigger, match=trigger)
+        else:
+            # if/else to avoid F811 "Redefinition of unused name" error in
+            # pyflakes 3.1.0+
+            execute_handler = handler
```

Comment:
This seems like a bug in `pyflakes` to me. The report is incorrect for our code, because `execute_handler` is definitely being used in this scope. Seems like [the upstream claims this is working as intended](https://github.com/PyCQA/pyflakes/issues/780). Although the example in that report does not use the redefined name, adding a statement that _does_ use it causes the same error.

I would rather that we leave it the way it was (which I find more legible) and add a `# noqa:F811` to the class definition line, maybe along with a remark about why we need to work around pyflakes's bug.

Reply:
I don't think it's a bug. If you keep following the issue chain, you'll eventually arrive at:

> a function overriding a variable is almost always an error -- if you want to do exceptional things you can utilize `# noqa` (you're already using flake8)

_Originally posted by @asottile in https://github.com/PyCQA/pyflakes/issues/760#issuecomment-1380461009_

… which matches what triggered the F811 error here: a `def` overriding a regular assignment to the same name that is _not used between assignment and redefinition_. That's not to say I'm exactly _happy_ about upstream updating behavior to detect this case as an error, but I think it _is_ doing what they intended it to do.

If we could arrive at a sensible spelling of something like the below (I guess most of what it needs is a good descriptive name for the intermediate `def` that isn't `execute_handler`), would you take that instead of a `# noqa`?

```py
# ...

execute_handler = handler
argspec = inspect.getfullargspec(handler)

if len(argspec.args) >= match_count:
    @functools.wraps(handler)
    def some_other_name(bot, trigger):
        return handler(bot, trigger, match=trigger)

    execute_handler = some_other_name
```

(I'd say moving the `execute_handler = handler` assignment closer to its redefinition is clearer regardless of the verdict.)

Reply:
Summarizing discussion from IRC: @dgw is right, this is not a bug in the tool, even though we both disagree with what the tool is checking for. The behavior is indeed specifically about redefinition using `def`, so using an intermediate name works. I'm still team `# noqa` but don't object to the name-swap workaround.

Reply:
please when quoting don't mention me into a thread. I already get enough notifications and I don't need the back and forth of some random PR in my inbox